### DISCUSSION TITLE
Exported tiddlers seems to have huge CSS overload!

### DISCUSSION THREAD

**Discussion by twMat** [TiddlyWiki/TiddlyWiki5#5439] [General]
Try this; create a new dummy tiddler and export it into HTML format. Then peek at the code. If the tiddler was exported from an empty TW, the code is 2300+ lines long. If the identical tiddler was exported via tiddlywiki.com then it is almost 3500 lines long. That difference in itself indicates something is not right. The file is 99% stylesheet of which **a mere fraction is actually used** for the remaining 1% HTML code!!!

I assume the main part is the vanilla stylesheet. The styles include stuff like "Error message and password prompt", "Plugin reload warning" and more that don't reasonably concern exported static tiddlers. It seems even sidebar stylings and editor stylings are included but there are typically no sidebar or editor in exported tids. And theres js-related styling stuff (e.g "drag'n drop") but the tiddlers are static so...

Is it possible to have the export engine run through the code to identify which classes are used and then **only include these** in the export? If that's not possible, then at export engine should at least exclude several classes via some modifiable "exclude list". By default, several of the things I bring up here above could be on such a list.

I think this would drastically shrink the output - right?

**Comment by pmario** [comment]
https://github.com/Jermolene/TiddlyWiki5/blob/2e47f277acfcc52ab872e84261385d254873f67e/core/ui/PageStylesheet.tid#L1-L15

As you can see, every tiddler tagged `$:/tags/Stylesheet` is exported, since the core doesn't do any optimization at all.

**Comment by pmario** [comment]
I think, the easiest way would be to export 2 files. The "small" HTML content and the "big" CSS. That's the way the "static" export mechanism works. 

The CSS would still be overkill, but it will be cacheable on the client side. So the browser would only need to load it once, if the caching settings are done right. 

CSS tree-shaking is a relatively complex task. You'll have to analyse the html code, find all the classes and then simulate the CSS dependency tree the site is using, to know which CSS definitions are active for any given html-element.

Then you have to find the CSS definitions, that are used per element above ... You basically have to simulate the behaviour of a browser.

**Comment by twMat** [comment]
I can see how the "tree shaking" would be complex but how a predefined whitelist with default style defs, that that the user can then edit to add to (or delete from). The whitelist would include the "obvious" parts from the default stylesheets.

The idea you mention to split it up into separate HTML and CSS files would 
1. still not solve the problem, would it?
2. two files is... not a single file. So I don't think it can be emailed and opened easily by anyone then, right?

**Reply by pmario** [reply]
The 1st point for me personally is, that

  1.  I'd export the tiddler and the CSS file as 2 files. 
  2. Then I'd run webpack with a CSS tree-shaking package, that would give me back 1 html file, that contains the "optimized" CSS. 

The problem is that point 2 will need a complete node development environment. 

BUT it would be a 100 times less work than implementing something half-backed into TW, implementing the optimisation from scratch.

The 2nd point is, .. I think many if not most users I know send mega-byte and mega-byte of cat-videos with whatsApp. .. 

So do we really need to care about a 400kByte (HelloThere) html-page, which can be zipped.

**Reply by twMat** [reply]
If your use case is for some static *website* or other "ambitious project" then, yes, I can see how it can make sense to post bake it like that. But for a quick'n dirty "export and email" then it's overkill. And I just don't like *that much* excess dead wood, let alone cat videos ;-)

Still, a combo of an editable whitelist (styles for tiddler view mode, wiki background, ...) *and* an editable blacklist (styles for edit modes, sidebar, ctrlpanel stuff, JS-stuff, ...) just feels like it would be a very transparent and *simple* solution, particularly to implement. Yes, some dead wood would slip through but not *all* stylesheets.

**Comment by twMat** [comment]
Could there be an editable code preview of the to-be-exported tids?
If there could, then something like Codemirror could perhaps make it easier to cut out undesired parts.

**Comment by AnthonyMuscio** [comment]
Do I understand that the exported static files would be the number of tiddler + 1 CSS file without this optimisation step?

So the more you publish the smaller the ratio of CSS file to the total
This seems an acceptable approach if the CSS file is not excessive.

Tony

**Reply by twMat** [reply]
Yes. The problem is just that the CSS is likely to be excessive in almost all cases since you typically only export a few tiddlers when using the HTML format. What you export is the open river tiddlers so it is just a few tiddlers. At least I assume that's how people user the HTML exporter. So the CSS is typically huge in comparison.

**Comment by pmario** [comment]
There have been some discussions, to split the main CSS into different parts. If this is done right, it would help single tiddler exports.

**Reply by twMat** [reply]
Yeah, I was thinking about that too. Still, as you note, the current HTML export exports *all* stylesheets so already now it'd make sense if we could control which ones to not include.

**Comment by twMat** [comment]
@pmario 

I found this semi-relevant [video](https://www.youtube.com/watch?v=ZSSqsEVsChc). It refers to a [CSS inliner tool](https://automattic.github.io/juice/). My interpretation is that this converts all `class="foo"` into `style="whatever foo contains"`.

Could the Export as HTML do this? Or is that again shaking the full tree?
